api/migrations/2020-11-08-174607_trigger_users/up.sql

```diff
@@ -0,0 +1,35 @@
+CREATE OR REPLACE FUNCTION delete_dangling_program_records()
+    RETURNS trigger AS
+$$
+BEGIN
+    WITH pbs as (
+        SELECT min(cpu_cycles) as cpu_cycles, min(instructions) as instructions, min(registers_used) as registers_used, min(stacks_used) as stacks_used
+        FROM "user_program_records" as upr
+        WHERE upr.user_id = NEW.user_id 
+        AND upr.program_spec_id = NEW.program_spec_id
+    )
+    DELETE FROM "user_program_records" WHERE id in (
+        SELECT id
+        FROM "user_program_records" as upr, pbs
+        WHERE upr.user_id = NEW.user_id AND upr.program_spec_id = NEW.program_spec_id
+        AND upr.cpu_cycles > pbs.cpu_cycles
+        AND upr.instructions > pbs.instructions
+        AND upr.registers_used > pbs.registers_used
+        AND upr.stacks_used > pbs.stacks_used
+        AND upr.id NOT IN (
+            -- get all referenced rows so they wont be deleted
+            SELECT record_id FROM "user_programs" as up
+            WHERE up.user_id = NEW.user_id AND up.program_spec_id = NEW.program_spec_id
+        )
+        AND upr.id <> NEW.id
```

Comment:
IMO using table aliases makes it a bit harder to read (one more thing to remember), do you mind using the full table names instead?

Also could you add a few more comments just to clarify what each part of the WHERE clause is for?